### PR TITLE
drivers: flash: stm32g4: Flush caches after erase and write

### LINKED ISSUE #32221: Sporadic kernel panics on stm32g4 flash erase/writes
**Describe the bug**
When writing big data chunks to the flash of an STM32G4 SoC, sporadic kernel panics occur.
It seems that a cache flush is needed after each erase. 

**To Reproduce**
Build mcuboot for the nucleo_g474re and update several times.
Sometimes you will see a panic.

**Expected behavior**
Flash write and erase should work without panic

**Impact**
Firmware updates on stm32g4 series do not work

**Logs and console output**
````
E: ***** BUS FAULT *****
E:   Precise data bus error
E:   BFAR Address: 0xfab1f1a4
E: r0/a1:  0x0000045d  r1/a2:  0x00000000  r2/a3:  0xfab1f1a4
E: r3/a4:  0x20000208 r12/ip:  0xa0000000 r14/lr:  0x080043e1
E:  xpsr:  0x6100000f
E: Faulting instruction address (r15/pc): 0x08003d6a
E: >>> ZEPHYR FATAL ERROR 0: CPU exception on CPU 0
E: Fault during interrupt handling
E: Current thread: 0x20000208 (unknown)
E: Halting system
````

### PR DESCRIPTION
**Pull request by alexanderwachter**
Flush the caches after erasing.
This commit fixes sporadic kernel panics when writing big data chunks
to the flash. (data bus errors)

Fixes #32221

### CODE REVIEW

File: drivers/flash/flash_stm32g4x.c [hunk 1]
```diff
@@ -159,6 +182,8 @@ static int erase_page(const struct device *dev, unsigned int offset)
 	/* Wait for the BSY bit */
 	rc = flash_stm32_wait_flash_idle(dev);
 
+	flush_cache(regs);
```
Comment (ABOSTM) [inline drivers/flash/flash_stm32g4x.c:185]: According to STM32G4 Reference Manual, and I found that:
Chapter 3.3.7

> **Programming and caches**
> If a Flash memory write access concerns some data in the data cache, the Flash write
> access modifies the data in the Flash memory and the data in the cache.
> If an erase operation in Flash memory also concerns data in the data or instruction cache,
> you have to make sure that these data are rewritten before they are accessed during code
> execution. If this cannot be done safely, it is recommended to flush the caches

So flush is necessary for erase operation only.

**Review by ABOSTM — CHANGES_REQUESTED** [review]

File: drivers/flash/flash_stm32g4x.c (hunk 1 above)
Comment (alexanderwachter) [reply drivers/flash/flash_stm32g4x.c:185]: Thanks. Fixed.

**Review by nvlsianpu — COMMENTED** [review]
The Reference Manual stands:
>If an erase operation in Flash memory also concerns data in the data or instruction cache,
you have to make sure that these data are rewritten before they are accessed during code
execution. If this cannot be done safely, it is recommended to flush the caches by setting the
DCRST and ICRST bits in the Flash access control register (FLASH_ACR).

Therefore I would flush cashes only prior the erase operation.

This is just comment as I'm not part of STM team here.

File: drivers/flash/flash_stm32g4x.c (hunk 1 above)
Comment (erwango) [reply drivers/flash/flash_stm32g4x.c:185]: Is it ?

Comment (nvlsianpu) [reply drivers/flash/flash_stm32g4x.c:185]: > Therefore I would flush cashes only prior the erase operation.

**Review by ABOSTM — APPROVED** [review]
LGTM

File: drivers/flash/flash_stm32g4x.c (hunk 1 above)
Comment (erwango) [reply drivers/flash/flash_stm32g4x.c:185]: Oups, got errored by github.

**Review by ABOSTM — CHANGES_REQUESTED** [review]
Can you update commit message, and explain it is not necessary for write

**Review by erwango — APPROVED** [review]

**Comment by ABOSTM** [comment]
> The Reference Manual stands:
> 
> > If an erase operation in Flash memory also concerns data in the data or instruction cache,
> > you have to make sure that these data are rewritten before they are accessed during code
> > execution. If this cannot be done safely, it is recommended to flush the caches by setting the
> > DCRST and ICRST bits in the Flash access control register (FLASH_ACR).
> 
> Therefore I would flush cashes only prior the erase operation.
> 
> This is just comment as I'm not part of STM team here.

I think it doesn't matter because it is not standard cache but an ART accelerator (Adaptive real-time memory accelerator).
As soon as there is a write operation both flash and cache are updated. So there is no need for a real flush but rather a cache invalidation. This is what is done. And so we should probably spoke about cache invalidate instead of flush ... but this is the word used in Reference Manual.

**Review by ABOSTM — APPROVED** [review]
LGTM

**Pull request merged**